File: 백준/Gold/1753. 최단경로/최단경로.java
import java.util.*;
import java.io.*;

public class Main {
	
	private static BufferedReader br;
	private static BufferedWriter bw;
	private static StringTokenizer st;
	private static List<Edge>[] graph;
	private static boolean[] visited;
	private static int[] cost;
	
	
	static class Edge implements Comparable<Edge> {
		
		int node;
		int weight;
		
		public Edge(int node, int weight) {
			this.node = node;
			this.weight = weight;
		}
		
		@Override
		public int compareTo(Edge e) {
			
			return this.weight - e.weight;
		}
	}
	
	public static void main(String[] args) throws IOException {
		
		br = new BufferedReader(new InputStreamReader(System.in));
		bw = new BufferedWriter(new OutputStreamWriter(System.out));
		
		st = new StringTokenizer(br.readLine());
		int V = Integer.parseInt(st.nextToken());
		int E = Integer.parseInt(st.nextToken());
		
		st = new StringTokenizer(br.readLine());
		int K = Integer.parseInt(st.nextToken());
		
		graph = new ArrayList[V + 1];
		cost = new int[V + 1];
		visited = new boolean[V+1];
		
		Arrays.fill(cost, Integer.MAX_VALUE);
		
		for (int i = 1; i <= V; i++) {
			graph[i] = new ArrayList<>(); 
		}
		
		for (int i = 0; i < E; i++) {
			st = new StringTokenizer(br.readLine());
			int start = Integer.parseInt(st.nextToken());
			int end = Integer.parseInt(st.nextToken());
			int weight = Integer.parseInt(st.nextToken());
			
			graph[start].add(new Edge(end, weight));
		}
		
		djikstra(K);
		
		StringBuilder sb = new StringBuilder();
		
		for (int i = 1; i <= V; i++) {
			int c = cost[i];
			
			if (c == Integer.MAX_VALUE) {
				sb.append("INF");
				sb.append("\n");
			} else {
				sb.append(c);
				sb.append("\n");
			}
			
		}
		
		bw.write(sb.toString());
		bw.flush();
		bw.close();
		
	}
	
	private static void djikstra(int K) {
	
		PriorityQueue<Edge> queue = new PriorityQueue<>();
		queue.offer(new Edge(K, 0));
		cost[K] = 0;
		
		while (!queue.isEmpty()) {
			
			Edge edge = queue.poll();
			int node = edge.node;
			
			if (visited[node]) {
				
				continue;
			}
			
			visited[node] = true;
			
			for (Edge e : graph[node]) {
				if (cost[e.node] > cost[node] + e.weight) {
					cost[e.node] = cost[node] + e.weight;
					queue.offer(new Edge(e.node, cost[e.node]));
				}
			}
		}
		
	}
}

File: 백준/Gold/1753. 최단경로/최단경로.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "./example.txt";
const inputs = fs.readFileSync(filePath).toString().trim().split("\n");
const [V, E] = inputs.shift().split(" ").map(Number);
const K = Number(inputs.shift());
let graph = Array.from({length: V + 1}, () => []);
let visited = new Array(V + 1).fill(false);
let dist = new Array(V + 1).fill(Number.MAX_SAFE_INTEGER);

class PriorityQueue {

    constructor() {
        this.queue = [0];
    }

    compare(e1, e2) {
        return e1.weight - e2.weight;
    }

    isEmpty() {
        return this.queue.length <= 1;
    }

    offer(edge) {

        let idx = this.queue.length;

        while (idx > 1 && this.compare(edge, this.queue[Math.floor(idx / 2)]) <= 0) {
            this.queue[idx] = this.queue[Math.floor(idx / 2)];
            idx = Math.floor(idx / 2);
        }
        this.queue[idx] = edge;
    }

    poll() {

        if (this.queue.length <= 1) {
            throw new Error("NoSuchElementException");
        }

        let first = this.queue[1];
        let last = this.queue.pop();
        let size = this.queue.length - 1;

        if (size === 0) {
            return first;
        }

        let pIdx = 1;
        let cIdx = 2;

        while (cIdx <= size) {

            if (this.queue[cIdx + 1] !== undefined && this.compare(this.queue[cIdx], this.queue[cIdx + 1]) > 0) {
                cIdx++;
            }

            if (this.compare(last, this.queue[cIdx]) <= 0) {
                break;
            }
            this.queue[pIdx] = this.queue[cIdx];

            pIdx = cIdx;
            cIdx *= 2;
        }

        this.queue[pIdx] = last;
        return first;
    }


}

class Edge {

    constructor(node, weight) {
        this.node = node;
        this.weight = weight;
    }

}

main();

function main() {

    for (let input of inputs) {
        let [start, end, weight] = input.split(" ").map(Number);
        graph[start].push(new Edge(end, weight));
    }

    dijkstra();

    let answer = dist.slice(1).map(d => (d === Number.MAX_SAFE_INTEGER ? "INF" : d)).join("\n");
    console.log(answer);

}

function dijkstra() {

    let queue = new PriorityQueue();
    queue.offer(new Edge(K, 0));
    dist[K] = 0;

    while (!queue.isEmpty()) {

        let edge = queue.poll();
        let node = edge.node;

        if (visited[node]) {
            continue;
        }

        visited[node] = true;

        for (let e of graph[node]) {
            if (dist[e.node] > dist[node] + e.weight) {
                dist[e.node] = dist[node] + e.weight;
                queue.offer(new Edge(e.node, dist[e.node]));
            }
        }
    }
}
